drivers/gpio/gpio_eos_s3.c

```diff
@@ -0,0 +1,432 @@
+ * to "port1".
+ */
+static int find_pad(GPIOCfgTypeDef *cfg, int port_num)
+{
+	/* Find out which PAD will be used for given GPIO */
+	if (port_num == 0) {
+		switch (cfg->ucGpioNum) {
+		case GPIO_0:
+			cfg->xPadConf->ucPin = PAD_6;
+			cfg->xPadConf->ucFunc = PAD6_FUNC_SEL_GPIO_0;
+			break;
+		case GPIO_1:
+			cfg->xPadConf->ucPin = PAD_9;
+			cfg->xPadConf->ucFunc = PAD9_FUNC_SEL_GPIO_1;
+			break;
+		case GPIO_2:
+			cfg->xPadConf->ucPin = PAD_11;
+			cfg->xPadConf->ucFunc = PAD11_FUNC_SEL_GPIO_2;
+			break;
+		case GPIO_3:
+			cfg->xPadConf->ucPin = PAD_14;
+			cfg->xPadConf->ucFunc = PAD14_FUNC_SEL_GPIO_3;
+			break;
+		case GPIO_4:
+			cfg->xPadConf->ucPin = PAD_18;
+			cfg->xPadConf->ucFunc = PAD18_FUNC_SEL_GPIO_4;
+			break;
+		case GPIO_5:
+			cfg->xPadConf->ucPin = PAD_21;
+			cfg->xPadConf->ucFunc = PAD21_FUNC_SEL_GPIO_5;
+			break;
+		case GPIO_6:
+			cfg->xPadConf->ucPin = PAD_22;
+			cfg->xPadConf->ucFunc = PAD22_FUNC_SEL_GPIO_6;
+			break;
+		case GPIO_7:
+			cfg->xPadConf->ucPin = PAD_23;
+			cfg->xPadConf->ucFunc = PAD23_FUNC_SEL_GPIO_7;
+			break;
+		default:
+			return -EINVAL;
+		}
+	} else if (port_num == 1) {
+		switch (cfg->ucGpioNum) {
+		case GPIO_0:
+			cfg->xPadConf->ucPin = PAD_24;
+			cfg->xPadConf->ucFunc = PAD24_FUNC_SEL_GPIO_0;
+			break;
+		case GPIO_1:
+			cfg->xPadConf->ucPin = PAD_26;
+			cfg->xPadConf->ucFunc = PAD26_FUNC_SEL_GPIO_1;
+			break;
+		case GPIO_2:
+			cfg->xPadConf->ucPin = PAD_28;
+			cfg->xPadConf->ucFunc = PAD28_FUNC_SEL_GPIO_2;
+			break;
+		case GPIO_3:
+			cfg->xPadConf->ucPin = PAD_30;
+			cfg->xPadConf->ucFunc = PAD30_FUNC_SEL_GPIO_3;
+			break;
+		case GPIO_4:
+			cfg->xPadConf->ucPin = PAD_31;
+			cfg->xPadConf->ucFunc = PAD31_FUNC_SEL_GPIO_4;
+			break;
+		case GPIO_5:
+			cfg->xPadConf->ucPin = PAD_36;
+			cfg->xPadConf->ucFunc = PAD36_FUNC_SEL_GPIO_5;
+			break;
+		case GPIO_6:
+			cfg->xPadConf->ucPin = PAD_38;
+			cfg->xPadConf->ucFunc = PAD38_FUNC_SEL_GPIO_6;
+			break;
+		case GPIO_7:
+			cfg->xPadConf->ucPin = PAD_45;
+			cfg->xPadConf->ucFunc = PAD45_FUNC_SEL_GPIO_7;
+			break;
+		default:
+			return -EINVAL;
+		}
+	} else {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/*
+ * NOTE:
+ * Pin number defined in the device tree should be actually a GPIO number
+ * to avoid overflow in high level GPIO API.
+ * `gpio_num_extended` in range of 0 to 7 matches to `port_num == 0` PADs from
+ * `find_pad` function and `gpio_num_extended` in range of 8 to 15
+ * are associated with `port_num == 1` consequently.
+ */
+static int gpio_eos_s3_configure(const struct device *dev,
+				 gpio_pin_t gpio_num_extended,
+				 gpio_flags_t flags)
+{
+	ARG_UNUSED(dev);
+	int ret;
+	GPIOCfgTypeDef gpio_cfg;
+	PadConfig pad_conf;
+	int port_num = 0;
+
+	gpio_cfg.xPadConf = &pad_conf;
+
+	/* Check for an invalid gpio number */
+	if (gpio_num_extended >= MAX_GPIOS) {
+		return -EINVAL;
+	}
+
+	/* Determine "port" number */
+	if ((int)(gpio_num_extended / MAX_GPIOS_PER_PORT) == 1) {
```

Comment:
I don't think the cast here should be necessary, if `MAX_GPIOS_PER_PORT` is unsigned (and, optionally, the RHS is `1U`).